clients/src/main/java/org/apache/kafka/common/requests/UpdateMetadataRequest.java

```diff
@@ -283,6 +329,25 @@ protected Struct toStruct() {
         }
     }
 
+    // federation
+    public void rewriteRemoteRequest(String routingClusterId, int controllerId, int controllerEpoch, long maxBrokerEpoch) {
+        // FIXME?  should we add a version check for 7+?  federation should not be enabled with less than that...
+        structLock.lock();
+        try {
+            data.setRoutingClusterId(routingClusterId);
+            data.setControllerId(controllerId);
+            data.setControllerEpoch(controllerEpoch);
+            // brokerEpoch apparently gets rewritten by the controller for every receiving broker (somewhere...):
```

Comment:
Thanks for the link, but that's how it's used after the request is received.  My comment was more about wherever it gets modified (in the request) prior to getting sent to any given broker, since broker epochs are per-broker, right?  Or am I misunderstanding something fundamental?  Is "broker epoch" not actually about brokers?  (There's a separate controller epoch, so I didn't think that was the case.)

As to the lock, the method is only setting four simple fields in data and clearing the struct variable, so it's not like the lock will be held for long in any case.  Is it really worth the risk?  Upstream code could change at some point, and then there would be a hard-to-find race condition lurking.  Seems a tad dangerous to me, don't you think?